sprint 2 finish

daaleche opened the pull request.

daaleche committed:
sprint 2 finish

mikebars1995 commented on src/utils/api.jsx:
```diff
@@ -6,7 +6,18 @@ export const checkReponse = (res) => {
         : res.json().then((err) => Promise.reject(err));
```
**Отлично**


Отличная функция проверки ответа от сервера

mikebars1995 commented on src/utils/api.jsx:
```diff
@@ -6,7 +6,18 @@ export const checkReponse = (res) => {
-export const getIngredients = () => {
+export const getIngredientsRequest = () => {
     return fetch(`${BASE_URL}/ingredients`)
         .then(checkReponse)
```
**Можно лучше**

Можно сделать **универсальную функцию запроса с проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(checkResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request` и убрать дублирование проверки на `ok`. Все остальное будет без изменений. Код станет чище

mikebars1995 commented on src/services/reducers/burger-constructor.jsx:
```diff
@@ -0,0 +1,54 @@
+            else
+                return {
+                    ...state,
+                    ingredients: [...state.ingredients, { ...action.ingredient, uuid: new_uuid() }],
```
**Нужно исправить**

Редьюсер должен быть **чистой функцией**. Генерировать `uuid` в самом редьюсере **нельзя**. Это делает редьюсер нечистой функцией. Нужно генерировать uuid при создании экшена (желательно, в **`action creator`**). И созданный uuid должен приходить в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста, если не получится
```js

// вот это - `action creator`
export const addIngridient = (item) => {
    return {
        type: ADD_INGRIDIENT,
        payload: {
            ...item, // используем `spread`, чтобы поменять ссылку на объект. Таким образом `redux` обновит его в хранилище
           uniqueId: uuid4()  // и добавляем в объект новое поле, которое потом будет использовано в `key`
        }
    }
}


```

А если используете `createSlice`, то это делается в `prepare`

```ts

prepare: (ingredient: TIngredient) => {
        return { payload: { ...ingredient, uniqueId: uuid4() } };
      }


```

Подробнее про `prepare` тут https://redux-toolkit.js.org/api/createSlice#customizing-generated-action-creators

mikebars1995 reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на *гитхабе* все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная


Что сделано отлично:


* Отличная функция проверки ответа от сервера

Что нужно исправить:

* Проект не запускается. Скрин https://disk.yandex.ru/i/v_nXfUAUuM909w
* Редьюсер должен быть **чистой функцией**. Генерировать `uuid` в самом редьюсере **нельзя**. Это делает редьюсер нечистой функцией. Нужно генерировать uuid при создании экшена (желательно, в **`action creator`**). И созданный uuid должен приходить в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста, если не получится


Что можно улучшить:

* Можно сделать **универсальную функцию запроса с проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе



Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

**Комментарии `Можно лучше` не обязательны к исправлению прямо сейчас, это рекомендации**

Удачного рефакторинга кода.

daaleche committed:
Изменения:
- в проект добавлены библиотеки, теперь проект должен запускаться
- исправлена функция редьюсера
- модернизация функции отправки запроса

mikebars1995 reviewed (commented):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.



**Можно лучше**

* Желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты
* лучше отображать **лоадер** (загрузку) при ожидании ответа от сервера с номером заказа
